### ISSUE TITLE
2.4: Leading zeros in `.if` interpreted as `false`

### ISSUE THREAD

**Issue by TimWolla** [haproxy/haproxy#1125]
## Detailed description of the problem

The `.if` in the config parser interprets numbers with leading zeros as false.

## Expected behavior

It should interpret that as `true` (or reject the configuration).

## Steps to reproduce the behavior

1. `./haproxy -c -f ./haproxy.cfg`

## Do you have any idea what may have caused this?

The `false` branch is taken when the first character of the argument is `0`.

## Do you have an idea how to solve the issue?

It should check that the argument is exactly `"0"`.

## What is your configuration?

```
.if 01
	.alert true
.else
	.alert false
.endif
```

## Output of `haproxy -vv` and `uname -a`

```
HA-Proxy version 2.4-dev8 2021/02/13 - https://haproxy.org/
Status: development branch - not safe for use in production.
Known bugs: https://github.com/haproxy/haproxy/issues?q=is:issue+is:open
Running on: Linux 4.4.0-201-generic #233-Ubuntu SMP Thu Jan 14 06:10:28 UTC 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits
  OPTIONS = 
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO -OPENSSL -LUA +FUTEX +ACCEPT4 -CLOSEFROM -ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS -OT -QUIC

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=4).
Built with network namespace support.
Built without compression support (neither USE_ZLIB nor USE_SLZ are set).
Compression algorithms supported : identity("identity")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built without PCRE or PCRE2 support (using libc's regex instead)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 5.4.0 20160609

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : none

Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace
```

## If HAProxy crashed: Last outputs and backtraces

n/a

## Additional information (if helpful)

n/a

**Comment by wtarreau** [comment]
Not even surprised to be honest :-)  I really wanted to "secure" the .if introduction and improve it later with better expression processors. I'd love to implement && || and even parenthesis though we'll probably start smaller.
At least supporting ints (correctly) and a few functions should be a good start for 2.4.

**Comment by TimWolla** [comment]
>  At least supporting ints (correctly) and a few functions should be a good start for 2.4.

In fact I wanted to look into implementing a few functions I'd consider useful to me, but then I stumbled upon the two issues I reported and wanted to get them out of the way first :smiley: To be clear, I don't plan to send a patch to fix them.

**Comment by wtarreau** [comment]
OK! Let's say we don't care for now, we can start a discussion on the list about the various functions that would be useful to have. Probably others will have good ideas as well.

**Comment by TimWolla** [comment]
This appears to be fixed since 6a2110c7175a031ff60b9b9769e80c3180ce0f9b.
